File: _skels/python-fastapi-rag-skel/app/wrapper_api/categories.py
"""`/api/categories` CRUD for the wrapper-shared categories table.

Categories are shared (not per-user) but all endpoints still require a
JWT bearer token so unauthenticated callers cannot modify them.
"""

from __future__ import annotations

from datetime import datetime
from typing import List

from fastapi import APIRouter, HTTPException, Response, status
from sqlalchemy.exc import IntegrityError
from sqlmodel import select

from .db import Category
from .deps import CurrentUser, SessionDep
from .schemas import CategoryCreate, CategoryRead

router = APIRouter(prefix="/api/categories", tags=["wrapper-categories"])


def _to_read(cat: Category) -> CategoryRead:
    return CategoryRead(
        id=cat.id,
        name=cat.name,
        description=cat.description or "",
        created_at=cat.created_at,
        updated_at=cat.updated_at,
    )


@router.get("", response_model=List[CategoryRead])
def list_categories(
    session: SessionDep,
    _user: CurrentUser,
) -> List[CategoryRead]:
    cats = session.exec(select(Category).order_by(Category.id)).all()
    return [_to_read(c) for c in cats]


@router.post(
    "",
    response_model=CategoryRead,
    status_code=status.HTTP_201_CREATED,
)
def create_category(
    payload: CategoryCreate,
    session: SessionDep,
    _user: CurrentUser,
) -> CategoryRead:
    cat = Category(
        name=payload.name,
        description=payload.description or "",
    )
    session.add(cat)
    try:
        session.commit()
    except IntegrityError:
        session.rollback()
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Category with name '{payload.name}' already exists",
        )
    session.refresh(cat)
    return _to_read(cat)


@router.get("/{category_id}", response_model=CategoryRead)
def retrieve_category(
    category_id: int,
    session: SessionDep,
    _user: CurrentUser,
) -> CategoryRead:
    cat = session.get(Category, category_id)
    if cat is None:
        raise HTTPException(status_code=404, detail="Category not found")
    return _to_read(cat)


@router.put("/{category_id}", response_model=CategoryRead)
def update_category(
    category_id: int,
    payload: CategoryCreate,
    session: SessionDep,
    _user: CurrentUser,
) -> CategoryRead:
    cat = session.get(Category, category_id)
    if cat is None:
        raise HTTPException(status_code=404, detail="Category not found")
    cat.name = payload.name
    cat.description = payload.description or ""
    cat.updated_at = datetime.utcnow()
    session.add(cat)
    try:
        session.commit()
    except IntegrityError:
        session.rollback()
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Category with name '{payload.name}' already exists",
        )
    session.refresh(cat)
    return _to_read(cat)


@router.delete("/{category_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_category(
    category_id: int,
    session: SessionDep,
    _user: CurrentUser,
) -> Response:
    cat = session.get(Category, category_id)
    if cat is None:
        raise HTTPException(status_code=404, detail="Category not found")
    session.delete(cat)
    session.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: _skels/rust-axum-skel/src/handlers/categories.rs
//! `/api/categories` CRUD — wrapper-shared category resource.
//!
//! Categories provide optional grouping for items via the
//! `items.category_id` FK. The schema mirrors the django-bolt skel's
//! `Category` model (table `categories`) so the shared
//! `_shared/db.sqlite3` is fully interchangeable between backends.
//! Every endpoint requires a Bearer JWT — anonymous requests get 401
//! from the `AuthUser` extractor.

use std::sync::Arc;

use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use chrono::Utc;
use sea_orm::sea_query::OnConflict;
use sea_orm::{
    ActiveModelTrait, DbErr, EntityTrait, ModelTrait, QueryOrder, RuntimeErr, Set,
};
use serde::Deserialize;

use crate::auth::AuthUser;
use crate::entities::category;
use crate::error::ApiError;
use crate::AppState;

#[derive(Debug, Deserialize)]
pub struct CreateCategoryPayload {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
}

pub async fn list_categories(
    State(state): State<Arc<AppState>>,
    _user: AuthUser,
) -> Result<impl IntoResponse, ApiError> {
    let rows = category::Entity::find()
        .order_by_asc(category::Column::Name)
        .all(&state.db)
        .await?;
    Ok(Json(rows))
}

pub async fn create_category(
    State(state): State<Arc<AppState>>,
    _user: AuthUser,
    Json(payload): Json<CreateCategoryPayload>,
) -> Result<impl IntoResponse, ApiError> {
    if payload.name.trim().is_empty() {
        return Err(ApiError::Validation(
            "category name cannot be empty".to_string(),
        ));
    }
    let now = Utc::now();
    let new_cat = category::ActiveModel {
        name: Set(payload.name.clone()),
        description: Set(payload.description),
        created_at: Set(now),
        updated_at: Set(now),
        ..Default::default()
    };
    let inserted = new_cat.insert(&state.db).await.map_err(|e| {
        if is_unique_violation(&e) {
            ApiError::Conflict(format!("category '{}' already exists", payload.name))
        } else {
            ApiError::Database(e)
        }
    })?;
    Ok((StatusCode::CREATED, Json(inserted)))
}

pub async fn get_category(
    State(state): State<Arc<AppState>>,
    _user: AuthUser,
    Path(id): Path<i32>,
) -> Result<impl IntoResponse, ApiError> {
    let row = category::Entity::find_by_id(id).one(&state.db).await?;
    let cat = row.ok_or_else(|| ApiError::NotFound(format!("category {id} not found")))?;
    Ok(Json(cat))
}

pub async fn update_category(
    State(state): State<Arc<AppState>>,
    _user: AuthUser,
    Path(id): Path<i32>,
    Json(payload): Json<CreateCategoryPayload>,
) -> Result<impl IntoResponse, ApiError> {
    if payload.name.trim().is_empty() {
        return Err(ApiError::Validation(
            "category name cannot be empty".to_string(),
        ));
    }
    let row = category::Entity::find_by_id(id).one(&state.db).await?;
    let existing = row.ok_or_else(|| ApiError::NotFound(format!("category {id} not found")))?;

    let mut active: category::ActiveModel = existing.into();
    active.name = Set(payload.name.clone());
    active.description = Set(payload.description);
    active.updated_at = Set(Utc::now());
    let updated = active.update(&state.db).await.map_err(|e| {
        if is_unique_violation(&e) {
            ApiError::Conflict(format!("category '{}' already exists", payload.name))
        } else {
            ApiError::Database(e)
        }
    })?;
    Ok(Json(updated))
}

pub async fn delete_category(
    State(state): State<Arc<AppState>>,
    _user: AuthUser,
    Path(id): Path<i32>,
) -> Result<impl IntoResponse, ApiError> {
    let row = category::Entity::find_by_id(id).one(&state.db).await?;
    let existing = row.ok_or_else(|| ApiError::NotFound(format!("category {id} not found")))?;
    existing.delete(&state.db).await?;
    Ok(StatusCode::NO_CONTENT)
}

/// Detect a UNIQUE-constraint violation across SQLite and Postgres
/// without depending on driver-specific error types.
fn is_unique_violation(err: &DbErr) -> bool {
    let msg = match err {
        DbErr::Query(RuntimeErr::SqlxError(sqlx_err)) => sqlx_err.to_string(),
        DbErr::Exec(RuntimeErr::SqlxError(sqlx_err)) => sqlx_err.to_string(),
        other => other.to_string(),
    };
    msg.contains("UNIQUE constraint failed")
        || msg.contains("duplicate key value")
        || msg.contains("violates unique constraint")
}

// Tag the (currently-unused) `OnConflict` import so future SeaORM
// upserts can reach for it without a fresh `use` line.
#[allow(dead_code)]
fn _on_conflict_handle() -> OnConflict {
    OnConflict::new()
}
